darwin bootstrap-tools bump LLVM to 11

toonn opened the pull request:
###### Motivation for this change

To bump the Apple SDK we need an updated bootstrap-tools for the Darwin stdenv.

###### Current status

I have added two tags on my branch. `bs1` points to the latest commit that can be built with the current bootstrap-tools, for everything after this you have to pass in this bootstrap-tools, and we need another bootstrap-tools to get `darwin.Csu` a dependency of stdenv to build, this would be tagged `bs2`. `current` points to the commit I'm currently working on, something's depending on a libunwind path it shouldn't be when trying to make-bootstrap-tools. The commits dealing with `darwin.xnu` and `darwin-stubs` are specific to the Apple SDK bump. I have a [messy build log](https://git.io/JmxTi) of this, which may help (or hinder) understanding the changes.

One big change in this branch is the bump from LLVM 7 to 11. This is because of [this build failure](https://git.io/J3T9N). I didn't really see a way forward and another part of `darwin.CF` was problematic because of LLVM 7 anyway so @LnL7 figured I might as well try bumping stdenv (and bootstrap-tools) LLVM, this was based on previous work trying to bump it to 9 #85151. If the priority is to bump only the Apple SDK (and we find an alternative solution to that build failure) it might be less time consuming to delay the LLVM bump and continue on my `apple_sdk-bump` branch.

cc @domenkozar

toonn committed:
bootstrap-tools: Update to LLVM 11

toonn committed:
bootstrap-tools: Fix xnu python3 patch

This patch only affects the xnu included in the bootstrap-tools to avoid
a stdenv rebuild.

toonn committed:
bootstrap-tools: Include libclang-cpp.dylib

Credits to @LnL7, this is based on commit
ae68f1c447d775e64d6e16a33f2affe591c419bf from PR #85151.

toonn committed:
bootstrap-tools: Include libclang-cpp.11.1.dylib too

We need to include the library with the version suffix because
`libclang-cpp.dylib` is just a symlink.

toonn committed:
bootstrap-tools: Add bootstrapFiles argument

I was having a hard time testing new bootstrapFiles because
`make-bootstrap-tools.nix` imports `pkgspath` but does not pass anything
but the current system.

This is merely for convenience and I'm not entirely certain it's a
sensible thing to do, maybe generating new bootstrapFiles while
overriding the current bootstrapFiles isn't something you're supposed to
do?

toonn committed:
bootstrap-tools: Patch TargetConditionals.h

`TargetConditionals.h` was missing several definitions, like
`TARGET_OS_TV` that are part of SDK 10.12 at least. And one that doesn't
seem to occur in any SDK afaict, `TARGET_OS_EMBEDDED_OTHER`.

I added the definitions from SDK 10.12 verbatim and defined
`TARGET_OS_EMBEDDED_OTHER` to be equal to `0`.

This is a modified version of a patch to avoid a stdenv rebuild.

SuperSandro2000 commented on pkgs/os-specific/darwin/darwin-stubs/default.nix:
```diff
@@ -2,12 +2,9 @@
-    url = "https://github.com/NixOS/darwin-stubs/releases/download/v20201216/10.12.tar.gz";
-    sha256 = "1fyd3xig7brkzlzp0ql7vyfj5sp8iy56kgp548mvicqdyw92adgm";
-  };
+  src = ~/src/darwin-stubs/stubs/10.13;
```
That won't work.

SuperSandro2000 commented on pkgs/os-specific/darwin/swift-corelibs/corefoundation.nix:
```diff
@@ -71,24 +73,7 @@ stdenv.mkDerivation {
-        continue
-    done
-  '';
+  buildPhase = "ninja -j $NIX_BUILD_CORES";
```
```suggestion
  buildPhase = ''
    runHook preBuild

    ninja -j $NIX_BUILD_CORES

    runHook postBuild
  '';
```

SuperSandro2000 commented on pkgs/os-specific/darwin/apple-source-releases/xnu/3789_70_16/default.nix:
```diff
@@ -0,0 +1,141 @@
+    mv $out/internal_hdr/include/mach/*.h $out/include/mach
+
+    # Get rid of some junk lying around
+    rm -rf $out/internal_hdr $out/usr $out/local
```
```suggestion
    rm -r $out/internal_hdr $out/usr $out/local
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-macos-monthly/12330/2

sternenseemann commented on pkgs/stdenv/darwin/make-bootstrap-tools.nix:
```diff
@@ -79,6 +87,7 @@ in rec {
 
       cp -rL ${llvmPackages.clang-unwrapped}/lib/clang $out/lib
 
+      cp -d ${llvmPackages.libclang}/lib/libclang-cpp*.dylib $out/lib
       cp -d ${llvmPackages.libcxx}/lib/libc++*.dylib $out/lib
       cp -d ${llvmPackages.libcxxabi}/lib/libc++abi*.dylib $out/lib
       cp -d ${llvmPackages.llvm.lib}/lib/libLLVM.dylib $out/lib
```
Since 7869d1654517c028aa76fc1dedc9b5ac301a867a something like this would be required:

```suggestion
      cp -rL ${llvmPackages.libclang.lib}/lib/clang $out/lib

      cp -d ${llvmPackages.libclang.lib}/lib/libclang-cpp*.dylib $out/lib
      cp -d ${llvmPackages.libcxx.lib}/lib/libc++*.dylib $out/lib
      cp -d ${llvmPackages.libcxxabi.lib}/lib/libc++abi*.dylib $out/lib
      cp -d ${llvmPackages.llvm.lib}/lib/libLLVM.dylib $out/lib
```

Mic92 replied:
`lib.getLib` is more future proof: `${lib.getLib llvmPackages.libcxxabi}`

toonn replied:
I've applied the suggestion everywhere a `/lib` is used. Thanks!

toonn committed:
bootstrap-tools: libclang-cpp already included

Changes to llvmPackages have caused the `libclang-cpp*.dylib` files to
be included in the `clang-unwrapped.lib` output. So we no longer need to
copy them from libclang.

toonn commented:
Ok, so the latest force-push is the branch with all the changes bumping from LLVM 9 to 11 for both the Darwin bootstrap-tools and the stdenv. 
To generate the bootstrap-tools for the tip of this branch (tag `bs2`) you need to first build an intermediate bootstrap-tools at tag `bs1`.

Commit 7db1708e adds an argument to `make-bootstrap-tools.nix` that can be used to do so. I suppose this isn't useful for hydra, we'll probably have to have it build a bootstrap at `bs1` and upload that somewhere and add a commit changing the url in stdenv?

For development I usually do the following:

```
> git checkout bs1
> nix-build pkgs/stdenv/darwin/make-bootstrap-tools.nix -A build
/nix/store/amq6frjddx18mmycj1ng0kc2r8rk4rbg-stdenv-bootstrap-tools
> git checkout bs2
> nix-build --no-out-link -E 'let np = import <nixpkgs> {}; bts = _: tool: "/nix/store/amq6frjddx18mmycj1ng0kc2r8rk4rbg-stdenv-bootstrap-tools/on-server/" + tool; bootstrapFiles = { bootstrapFiles = np.lib.mapAttrs bts { bzip2 = "bzip2"; cpio = "cpio"; mkdir = "mkdir"; sh = "sh"; tarball = "bootstrap-tools.cpio.bz2"; }; }; cfg = { stdenvStages = args: let args_ = args // bootstrapFiles; in (import pkgs/stdenv/darwin args_).stagesDarwin; }; make-bootstrap-tools = import pkgs/stdenv/darwin/make-bootstrap-tools.nix bootstrapFiles; pkgs = import ./. cfg; in make-bootstrap-tools.build'
```

This relies on two commits since merging the Apple Silicon PR #105026, 2987d66d and 38207735. This may be relevant if we target staging. The latter wasn't present there last I checked.

domenkozar commented:
@toonn if this is ready for review, mark this PR as not draft.

toonn commented:
@domenkozar, this PR is still a draft. There's probably package breakages we need to fix, working with @LnL7 to get a hydra jobset for this.

I'm a little stumped by the ofborg-eval error tbh. I didn't touch `llvm-manpages`, it's still defined for LLVM 11 and it's still in `allpackages.nix`. I'm not sure what the attribute being missing can mean at this point.

domenkozar commented:
That error will be fixed with https://github.com/NixOS/nixpkgs/pull/125184, hope to merge that tomorrow

toonn commented:
@domenkozar, ah, thanks. I had the weirdest case of déjá vu because of the "recommonmark." It's because I saw that PR earlier.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-macos-monthly/12330/4

LnL7 commented on pkgs/stdenv/darwin/make-bootstrap-tools.nix:
```diff
@@ -1,14 +1,20 @@
+let cross = if crossSystem != null
+      then { inherit crossSystem; }
+      else {};
+    custom-bootstrap = if bootstrapFiles != null
```
Is this just a convenience to test the stdenv, etc. using the new bootstrap-tools without having to edit `pkgs/stdenv/darwin/default.nix`?

toonn replied:
Yep, as I described [here](https://github.com/NixOS/nixpkgs/pull/121055#issuecomment-852429669).

LnL7 commented on pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix:
```diff
@@ -63,21 +63,28 @@ appleDerivation' stdenv {
-    #define TARGET_OS_IPHONE        0
-    #define TARGET_IPHONE_SIMULATOR 0
-    #define TARGET_OS_LINUX         0
+    #define TARGET_OS_MAC               1
```
Shouldn't be needed for the initial bootstrap build yet, removing the stdenv rebuild.

toonn replied:
You're right, forgot to remove it when adding it in `make-bootstrap-tools.nix`. >.<

toonn replied:
I'm still building the intermediate bootstrap-tools to verify that the preFixup hook actually works as expected btw.

toonn replied:
Overridden installPhase did the trick.

toonn replied:
Should I go ahead and mark this ready for review?

sternenseemann commented:
What happened to the global `llvmPackages` bump?

toonn commented:
@sternenseemann, that's upcoming. We need this intermediate bootstrap-tools built first.

toonn commented:
@sternenseemann, continuing in #126411.
